### ISSUE TITLE
Support Covariant Opaques (documentation)

### ISSUE THREAD

**Issue by mheiber** [erlang/otp#5240]
@Feature: support covariant opaques (documentation)

**Is your feature request related to a problem? Please describe.**

I think it's beneficial if the community has a shared understanding of what `-opaque` means. In particular, a variety of tools consume Erlang type annotations, and ideally the tools would be compatible.

I think the language ideally would have an opinion on what the variance of `T` in `-opaque foo(T) :: .....` is, in order for opaques to actually be opaque, based on [the definition in the docs](https://erlang.org/doc/reference_manual/typespec.html)

> "Types declared as opaque represent sets of terms whose structure is not supposed to be visible from outside of their defining module. That is, only the module defining them is allowed to depend on their term structure. "

Currently, any covariant uses of opaques are "depending on [the opaque's] term structure": if the definition changes such that the type is no longer covariant, then users' code breaks.

The best use case for parameterized opaques, imo, is covariant containers like `sets:set/1`, `maps:iterator/1`, `gb_sets:set/1`, etc. In practice all opaques are covariant: I have yet to see a non-covariant opaque. 

**Describe the solution you'd like**

- Document that all opaque definitions are assumed to be covariant. In practice, this just means that the docs discourage users from defining opaques where a type parameter appears in function type parameter position: `-opaque bad(T) :: fun((T) -> term()).
- Follow-on issue: support covariant opaques in Dialyzer. Using opaques as if they are covariant works most of the time in Dialyzer, but there are cases where there are spurious errors due to Dialyzer assuming opaques are invariant. [Example and discussion](https://github.com/erlang/otp/pull/5223#issuecomment-925696731)

**Describe alternatives you've considered**

1. Treat opaques as invariant by default, allow annotations for covariant and contravariant parameters

This idea was suggested by @jhogberg in [a comment on #5223](https://github.com/erlang/otp/pull/5223#issuecomment-924917745). I think it would look something like:

`-opaque foo(InvariantParam, +CovariantParam, -ContravariantParam)`

Requiring such variance annotations seems like a high burden for users: not just at the use site, but at the call site as well, since one has to look at the source code of the opaque type to see what the variance is in order to know whether one is using the type correctly. The burden would be worth it if Erlang were a mutation-heavy language with lots of compound data structures, but Erlang is not that kind of language and I haven't seen any contravariant opaques. Type variables are already [pretty difficult for some devs to understand](https://youtu.be/_u1NDuFsW2A?t=1528), and I worry variance annotations would make the problem worse.

> An open question is whether plain type aliases (`-type`) would also require annotations for variance.

2. Opaques covariant by default, allow annotations for invariant and contravariant parameters

This is consistent with my proposal and could be done as optional follow-on work.

**Additional context**

This issue grew out of discussion on https://github.com/erlang/otp/pull/5223

**Comment by jhogberg** [comment]
I discussed this with @bjorng earlier today, and I'm coming around to the idea of making covariant the default and adding annotations later on if needed. We'll need to write an EEP but implementing it will be trivial.

> Type variables are already [pretty difficult for some devs to understand](https://youtu.be/_u1NDuFsW2A?t=1528), and I worry variance annotations would make the problem worse.

The irony is that the presentation gets them wrong by having them make sense and be useful. As documented they're quite counterintuitive and hard to use correctly, and I would much prefer them to work as described in the video.

Since `dialyzer` ignores type variables in specs at the moment and we can't think of a tool that handles them as documented, we're going to try bringing them in line with how they work in other languages. We'd love to have you on board.

**Comment by mheiber** [comment]
> We'll need to write an EEP but implementing it will be trivial.

I'd be happy to collaborate! Thanks for your attention to this issue.

**Comment by mheiber** [comment]
cc @josefs

**Comment by mheiber** [comment]
I wrote up a brief guide below to opaques and recommendations for using opaques. If you think this can be adapted for OTP docs, I'd be happy to send a diff. I hope we can fill the gap in [the current documentation of opaques](https://www.erlang.org/doc/reference_manual/typespec.html#type-declarations-of-user-defined-types), which is accurate but very brief: "Types declared as opaque represent sets of terms whose structure is not supposed to be visible from outside of their defining module.".


## Understanding Opacity in Erlang

TLDR: The main use case for opacity in Erlang is to hide the implementation of a data type, enabling evolving the API while minimizing risk of breaking consumers. The runtime does not check opacity. Dialyzer provides some opacity-checking, but the rest is up to convention.

I'll explain what Erlang opacity is (and the trade-offs involved) via the example of OTP's `set()` data type. This type *was* defined in `sets` module like this:

```erl
-opaque set(Element) :: #set{segs :: segs(Element)}.
```

OTP 24  changed the definition to the following, in commit 7f6d31bef41ee923d5d1552a4b1fdacb975cb919:

```erl
-opaque set(Element) :: #set{segs :: segs(Element)} | #{Element => ?VALUE}.
```

At runtime, in OTP 24 sets are exclusively represented by maps rather than record/tuples.

and this change was safer and more backwards-compatible than if the type had been defined with `-type` instead of `-opaque`. Here's why: when a module defines an `-opaque`, the contract is that only the defining module should rely on the definition of the type: no other modules should rely on the definition.

This means that code that pattern-matched on `set` as a record/tuple technically broke the contract, and opted-in to being potentially broken when the definition of `set()` changed. Before OTP 24, this code printed `ok`. In OTP 24 it errors:

```erl
Set = sets:new(),
case Set of
    X when is_tuple(Set) ->
        io:format("ok")
end
```

**When working with an opaque defined in another module, here are some recommendations:**
- Don't examine the underlying type using pattern-matching, guards, or functions that reveal the type, such as `size/1`.
- Instead, use functions provided by the module for working with the type. For example, `sets` module provides `sets:new/0`, `sets:add/2`, `sets:is_element/2`, etc.
- You typically *can* rely on the property that type variables preserve the subtype relationship (covariance). This is easiest to see with an example: `sets:set(a)` is a subtype of `sets:set(a | b)` and not the other way around.

**When defining your own opaques, here are some recommendations:**
- Since consumers are expected to not rely on the definition of the opaque type, you must provide functions for constructing and querying/deconstructing intances of your opaque type. For example, sets can be constructed with `sets:new/0`, `sets:from_list/1`, `sets:add/2`, queried with `sets:is_element/2`, and deconstructed with `sets:to_list/1`.
- Don't define an opaque with a type variable in parameter position. This breaks the normal and expected behavior that (for example) `my_type(a)` is a subtype of `my_type(a | b)`

Note that opaques can be harder to work with for consumers, since the consumer is expected not to pattern-match. When providing an opaque type, a module should also typically provide functions that can create instances without breaking opacity (`like sets:new/0`) and use instances without breaking opacity (`like sets:add/2` and `sets:is_element/2`).

Also, opacity in Erlang is skin-deep: the runtime does not enforce opacity-checking. So now that sets are implemented in terms of maps, an `is_map` check on a set *will* pass. The opacity rules are only enforced by convention and by additional tooling such as Dialyzer. And this enforcement is not total: For example, determined consumer of `sets` can still do things that reveal the structure of the set, such as by printing, serializing, or using a set as `term()` and then inspecting via functions like `is_map` or `maps:get/2`. And Dialyzer must make some [approximations](https://github.com/erlang/otp/issues/5118). Opacity checking has limitations, but is still a vital tool in scalable Erlang development.

I hope guide serves as a useful reference for the trade-offs involved in using Erlang's opacity features, as well as what opacity does and doesn't provide.

**Comment by bjorng** [comment]
> Don't examine the underlying type using pattern-matching, guards, or functions that reveal the type, such as size/1.

I suggest you change `size/1` to `tuple_size/1`. While `size/1` is not deprecated (we don't dare deprecate it because too much code uses it), for new code it is better to use `tuple_size/1` or `byte_size/1`.

> You typically can rely on the property that type variables preserve the subtype relationship (covariance).

Could you phrase that in a clearer way? Perhaps it is better not to mention "covariance". Perhaps give the example first and give the more formal explanation after the example.

Otherwise, it looks good to me and I think it wold be an improvement of the documentation. @jhogberg?

**Comment by mheiber** [comment]
> I suggest you change `size/1` to `tuple_size/1`
Thanks, fixed

> Perhaps it is better not to mention "covariance".  Perhaps give the example first and give the more formal explanation after the example.
Thanks, I updated to lead with the set example and then explain the general property in a (hopefully) digestable way.

**Comment by josefs** [comment]
Sorry for being slow to reply to this thread.

Covariant opaques definitely seems like a useful feature. And making them covariant by default is certainly convenient. But what do we do with in/contra-variant types? Having only support for contravariance is unsound. There are two options here as I see it:

- Infer the variance of the type parameters. This is very convenient in the vast majority of cases. But having only inference means that a library author cannot declare their type e.g. invariant if they would want to.
- Allow annotations of type parameters and let the default be convariance for unannotated parameters. A typechecker would still  check that the annotations are correct. This option gives more control to library authors to declare variance, while at the same time remove noise in the vast majority of cases.

I prefer the second option.

Re the documentation for opaques @mheiber, it looks great!

Side note:
It's not a good idea to use the video of my presentation to argue that type variables are confusing for developers. Here's what actually happened in the room: there were two people that kept asking questions about variables, but the majority of the room seemed to get the topic. It doesn't come across well in the video, but there were a lot of people chiming in and trying to help these guys grok the concept.

**Comment by mheiber** [comment]
Another option is to simply not support non-covariant aliases. There are
none in OTP and I only saw one in the wild, ever, which was easily
refactored. This may be because the only source of contravariance in Erlang
is when a type alias that sticks a type argument into function parameter
position, which is pretty weird and
could surprise the caller.

If there is a killer use case for non-covariant opaques, @josefs's second
option sounds good to me, fwiw. @jhogberg <https://github.com/jhogberg>
suggested the same in a comment on #5223
<https://github.com/erlang/otp/pull/5223#issuecomment-924917745>, so this
seems like a case of great minds thinking alike.

Re docs: I should get around to doing this. I have to remember how to write
XML.
